google-cloud-bigquerystorage/src/test/java/com/google/cloud/bigquery/storage/v1alpha2/JsonToProtoMessageTest.java

```diff
@@ -0,0 +1,482 @@
+    String[] stringArr = {"hello", "this", "is", "a", "test"};
+    int[] intArr = {1, 2, 3, 4, 5};
+
+    JSONObject json = new JSONObject();
+    json.put("double", new JSONArray(doubleArr));
+    json.put("int", new JSONArray(intArr));
+    JSONObject jsonRepeatedString = new JSONObject();
+    jsonRepeatedString.put("test_repeated", new JSONArray(stringArr));
+    json.put("repeated_string", jsonRepeatedString);
+
+    DynamicMessage protoMsg =
+        JsonToProtoMessage.convertJsonToProtoMessage(NestedRepeated.getDescriptor(), json);
+    AreMatchingFieldsFilledIn(protoMsg, json);
+  }
+
+  @Test
+  public void testNestedRepeatedComplexFail() throws Exception {
+    double[] doubleArr = {1.1, 2.2, 3.3, 4.4, 5.5};
+    Boolean[] fakeStringArr = {true, false};
+    int[] intArr = {1, 2, 3, 4, 5};
+
+    JSONObject json = new JSONObject();
+    json.put("double", new JSONArray(doubleArr));
+    json.put("int", new JSONArray(intArr));
+    JSONObject jsonRepeatedString = new JSONObject();
+    jsonRepeatedString.put("test_repeated", new JSONArray(fakeStringArr));
+    json.put("repeated_string", jsonRepeatedString);
+
+    try {
+      DynamicMessage protoMsg =
+          JsonToProtoMessage.convertJsonToProtoMessage(NestedRepeated.getDescriptor(), json);
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          e.getMessage(),
+          "JSONObject does not have a string field at root.repeated_string.test_repeated[0].");
+    }
+  }
+
+  @Test
+  public void testAllowUnknownFields() throws Exception {
+    JSONObject json = new JSONObject();
+    json.put("test_repeated", new JSONArray(new int[] {1, 2, 3, 4, 5}));
+    json.put("string", "hello");
+
+    DynamicMessage protoMsg =
+        JsonToProtoMessage.convertJsonToProtoMessage(RepeatedInt64.getDescriptor(), json, true);
+    AreMatchingFieldsFilledIn(protoMsg, json);
+  }
+
+  @Test
+  public void testAllowUnknownFieldsError() throws Exception {
+    JSONObject json = new JSONObject();
+    json.put("double", 1.1);
+    json.put("string", "hello");
+
+    try {
+      DynamicMessage protoMsg =
+          JsonToProtoMessage.convertJsonToProtoMessage(RepeatedInt64.getDescriptor(), json);
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          e.getMessage(),
+          "JSONObject has unknown fields. Set allowUnknownFields to True to ignore unknown fields.");
+    }
+  }
+
+  @Test
+  public void testTopLevelMatchFail() throws Exception {
+    JSONObject json = new JSONObject();
+    json.put("double", 1.1);
+    try {
+      DynamicMessage protoMsg =
+          JsonToProtoMessage.convertJsonToProtoMessage(Int64Type.getDescriptor(), json);
+    } catch (IllegalArgumentException e) {
+      assertEquals(
+          e.getMessage(),
+          "There are no matching fields found for the JSONObject and the protocol buffer descriptor.");
+    }
+  }
+
+  @Test
+  public void testTopLevelMatchSecondLevelNoMatch() throws Exception {
```

Comment:
Should this be a failed case?

Reply:
This one should not be a failed case since the top level has matching fields (both test_int and complexLvl2) but the second level which is the optional field test_int in ComplexLvl2 did not match. However since the top level matched already, it doesn't matter if the second level had any matching fields. The json data is legal too since the test_int in ComplexLvl2 is optional.

Reply:
But is the no_match field counted as an unknown field?